src/session/content/chat_action_bar.rs

```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
         self.save_message_as_draft();
 
         let imp = self.imp();
+        let mut bindings = imp.bindings.borrow_mut();
+        while let Some(binding) = bindings.pop() {
+            binding.unwatch();
+        }
 
         if let Some(ref chat) = chat {
             self.load_draft_message(chat.draft_message());
 
             imp.chat_action_in_cooldown.set(false);
+
+            let chat_expression = gtk::ConstantExpression::new(chat).upcast();
+            let permissions_expression = gtk::PropertyExpression::new(
+                Chat::static_type(),
+                Some(chat_expression.clone()),
+                "permissions",
+            )
+            .upcast();
+
+            // Handle whether or not message bar should be shown
+            match chat.type_() {
+                ChatType::Private(data) => {
+                    let blocked_expression = gtk::PropertyExpression::new(
+                        Chat::static_type(),
+                        Some(gtk::ConstantExpression::new(chat)),
+                        "is-blocked",
+                    )
+                    .upcast();
+                    let user_expression = gtk::ConstantExpression::new(data).upcast();
+                    let visible_expression = gtk::ClosureExpression::with_callback(
+                        &[blocked_expression, user_expression],
+                        |args| {
+                            let is_blocked = args[1].get::<bool>().unwrap();
+                            let user = args[2].get::<User>().unwrap();
+                            if let UserType::Deleted = user.type_().0 {
+                                false
+                            } else {
+                                !is_blocked
+                            }
+                        },
+                    )
+                    .bind(&*imp.scrolled_window, "visible", Some(chat));
```

Comment:
Instead of doing the same binding (and also adding it to the bindings Vec) in every single variant, you can return the `visible_expression` of each variant and then doing the actual binding after the `match`.